tests/Unit/DataStructures/Tensor/Expressions/Test_Contract.cpp

```diff
@@ -0,0 +1,540 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/Contract.hpp"
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/Requires.hpp"
+
+namespace {
+template <typename... Ts>
+void create_tensor(gsl::not_null<Tensor<double, Ts...>*> tensor) noexcept {
+  std::iota(tensor->begin(), tensor->end(), 0.0);
+}
+
+template <typename... Ts>
+void create_tensor(gsl::not_null<Tensor<DataVector, Ts...>*> tensor) noexcept {
+  double value = 0.0;
+  for (auto index_it = tensor->begin(); index_it != tensor->end(); index_it++) {
+    for (auto vector_it = index_it->begin(); vector_it != index_it->end();
+         vector_it++) {
+      *vector_it = value;
+      value += 1.0;
+    }
+  }
+}
+
+template <typename DataType>
+void test_contractions_rank2(const DataType& used_for_size) noexcept {
+  // Contract (upper, lower) tensor
+  // Use explicit type (vs auto) so the compiler checks the return type of
+  // `evaluate`
+  Tensor<DataType, Symmetry<2, 1>,
+         index_list<SpatialIndex<3, UpLo::Up, Frame::Inertial>,
+                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
+      Rul(used_for_size);
+  create_tensor(make_not_null(&Rul));
+
+  const Tensor<DataType> RIi_contracted =
+      TensorExpressions::evaluate(Rul(ti_I, ti_i));
+
+  DataType expected_RIi_sum = make_with_value<DataType>(used_for_size, 0.0);
+  for (size_t i = 0; i < 3; i++) {
+    expected_RIi_sum += Rul.get(i, i);
+  }
+  CHECK(RIi_contracted.get() == expected_RIi_sum);
+
+  // Contract (lower, upper) tensor
+  Tensor<DataType, Symmetry<2, 1>,
+         index_list<SpacetimeIndex<3, UpLo::Lo, Frame::Grid>,
+                    SpacetimeIndex<3, UpLo::Up, Frame::Grid>>>
+      Rlu(used_for_size);
+  create_tensor(make_not_null(&Rlu));
+
+  const Tensor<DataType> RgG_contracted =
+      TensorExpressions::evaluate(Rlu(ti_g, ti_G));
+
+  DataType expected_RgG_sum = make_with_value<DataType>(used_for_size, 0.0);
+  for (size_t g = 0; g < 4; g++) {
+    expected_RgG_sum += Rlu.get(g, g);
+  }
+  CHECK(RgG_contracted.get() == expected_RgG_sum);
+}
+
+template <typename DataType>
+void test_contractions_rank3(const DataType& used_for_size) noexcept {
+  // Contract first and second indices of (upper, lower, lower) tensor
```

Comment:
Typo: it looks like you're testing the (lower, upper, lower) case?

Reply:
Done! (I really need to check my comments more carefully next time, thanks for catching these)